skmorf/tree/_unsup_criterion.pyx

```diff
@@ -133,35 +208,223 @@ cdef class TwoMeans(Criterion):
+
+        # Update statistics up to new_pos
+        #
+        # Given that
+        #   sum_left[x] +  sum_right[x] = sum_total[x]
+        # and that sum_total is known, we are going to update
+        # sum_left from the direction that require the least amount
+        # of computations, i.e. from pos to new_pos or from end to new_po.
+        if (new_pos - pos) <= (end - new_pos):
+            for p in range(pos, new_pos):
+                i = sample_indices[p]
+
+                if sample_weight is not None:
+                    w = sample_weight[i]
+
+                # TODO: sum_left updated
+                for k in range(self.n_outputs):
+                    self.sum_left[k] += w * self.X[i]
+
+                self.weighted_n_left += w
+        else:
+            self.reverse_reset()
+
+            for p in range(end - 1, new_pos - 1, -1):
+                i = sample_indices[p]
+
+                if sample_weight is not None:
+                    w = sample_weight[i]
+
+                # TODO: sum_left updated
+                for k in range(self.n_outputs):
+                    self.sum_left[k] -= w * self.X[i]
+
+                self.weighted_n_left -= w
+
+        # Update right part statistics
+        self.weighted_n_right = (self.weighted_n_node_samples - 
+                                self.weighted_n_left)
+        # TODO: no notion of classes
+        for k in range(self.n_outputs):
+            self.sum_right[k] = self.sum_total[k] - self.sum_left[k]
+
+        self.pos = new_pos
+        return 0
 
     cdef double node_impurity(self) nogil:
-        # need to implement
-        pass
+        """Evaluate the impurity of the current node.
+
+        Evaluate the TwoMeans criterion as impurity of the current node,
+        i.e. the impurity of sample_indices[start:end]. The smaller the impurity the
+        better.
+
+        TODO:
+        Review on TwoMeans specific methods: Good start, and it's def in the right direction.
+        impurity in TwoMeans is defined as the Variance. 
+        
+        Variance formula would be 
+        impurity = weight * (left_impurity + right_impurity) / n_samples 
+        and then left_impurity and right_impurity are defined as: 
+        weight * Variance of left child and weight * Variance of right child. 
+
+        So you need to compute variance in node_impurity and children_impurity. 
+        The only thing changing are the pointers.
+
+        You should therefore define a separate function to compute variance.
+        Focus on getting the Criterion correct and ping me w/ questions. We can move onto the splitter once this is done.
+        """
+
+        cdef double impurity = 0.0
+
+        # TODO: impurity in TwoMeans criterion is the variance.
+        # variance = weight * (left_impurity + right_impurity) / n_samples
+        impurity =  self._compute_variance() / self.weighted_n_node_samples
+
+        return impurity
+
+    cdef void _compute_variance(self) nogil:
```

Comment:
Variance function can be written in general as:

```
cdef double _compute_variance(const DTYPE_t[:] Xf,
            double mean,
            SIZE_t* samples,
            DOUBLE_t* sample_weight,
            SIZE_t start, SIZE_t end) nogil:
    """Compute sample variance from vector of feature values.

    Parameters
    ----------
    """
    cdef SIZE_t p
    cdef SIZE_t i
    cdef DOUBLE_t w = 1.0
    cdef double variance = 0.0
    cdef SIZE_t n_samples = end - start

    # compute sample variance
    for p in range(start, end):
        i = samples[p]

        # w is originally set to be 1.0, meaning that if no sample weights
        # are given, the default weight of each sample is 1.0
        if sample_weight != NULL:
            w = sample_weight[i]
        
        # ith sample, and pth feature
        variance += w * (Xf[i] - mean) * (Xf[i] - mean)
    variance /= n_samples
    return variance
```

Please look at this and see either draft an implementation for `children_impurity` and `node_impurity`, or write down question(s) for what you do not understand and why.

Reply:
Thank you for this :smiley: 

I'll push the code momentarily!